File: sdk/typescript/src/execution-engine.ts
/**
 * ExecutionEngine SDK client for on-chain verifiable computation.
 *
 * Wraps the ExecutionEngine.sol contract using viem, providing typed
 * methods for the full request lifecycle: submit, claim, prove, query.
 */
import type {
  PublicClient,
  WalletClient,
  GetContractReturnType,
} from 'viem';
import { getContract, parseEther } from 'viem';
import { executionEngineAbi } from './abi';

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

type Hex = `0x${string}`;

/** On-chain request status enum (mirrors ExecutionEngine.RequestStatus) */
export enum OnChainRequestStatus {
  Pending = 0,
  Claimed = 1,
  Completed = 2,
  Expired = 3,
  Cancelled = 4,
}

/** Decoded on-chain ExecutionRequest struct */
export interface OnChainExecutionRequest {
  id: bigint;
  imageId: Hex;
  inputDigest: Hex;
  requester: Hex;
  createdAt: number;
  expiresAt: number;
  callbackContract: Hex;
  status: OnChainRequestStatus;
  claimedBy: Hex;
  claimedAt: number;
  claimDeadline: number;
  tip: bigint;
  maxTip: bigint;
}

/** Options for {@link ExecutionEngineClient.submitRequest} */
export interface SubmitRequestOptions {
  /** Program image ID (bytes32) */
  imageId: Hex;
  /** Keccak256 hash of the inputs (bytes32) */
  inputDigest: Hex;
  /** URL where provers can fetch inputs (default: empty string) */
  inputUrl?: string;
  /** Contract that receives the callback (default: zero address) */
  callbackContract?: Hex;
  /** Seconds until the request expires (0 uses contract default of 1 hour) */
  expirationSeconds?: bigint;
  /** Input type: 0 = public, 1 = private (default: 0) */
  inputType?: number;
  /** Tip in ETH (e.g. '0.001'). Sent as msg.value. */
  tipEther: string;
}

/** On-chain prover statistics */
export interface OnChainProverStats {
  completed: bigint;
  earnings: bigint;
}

// ---------------------------------------------------------------------------
// Client
// ---------------------------------------------------------------------------

/**
 * Typed client for the ExecutionEngine smart contract.
 *
 * Read-only operations (getRequestStatus, getRequest, etc.) only require a
 * `PublicClient`. Write operations (submitRequest, claimJob, submitResult,
 * disputeResult) additionally require a `WalletClient`.
 *
 * @example
 * ```ts
 * import { createPublicClient, createWalletClient, http } from 'viem';
 * import { privateKeyToAccount } from 'viem/accounts';
 * import { sepolia } from 'viem/chains';
 * import { ExecutionEngineClient } from '@worldzk/sdk';
 *
 * const publicClient = createPublicClient({ chain: sepolia, transport: http() });
 * const walletClient = createWalletClient({
 *   chain: sepolia,
 *   transport: http(),
 *   account: privateKeyToAccount('0x...'),
 * });
 *
 * const engine = new ExecutionEngineClient(publicClient, walletClient, '0x...');
 *
 * const requestId = await engine.submitRequest({
 *   imageId: '0x1234...',
 *   inputDigest: '0xabcd...',
 *   tipEther: '0.001',
 * });
 * ```
 */
export class ExecutionEngineClient {
  private readonly publicClient: PublicClient;
  private readonly walletClient: WalletClient | undefined;
  private readonly contractAddress: Hex;

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  private readonly readContract: GetContractReturnType<typeof executionEngineAbi, any, Hex>;

  constructor(
    publicClient: PublicClient,
    walletClient: WalletClient | undefined,
    contractAddress: Hex,
  ) {
    this.publicClient = publicClient;
    this.walletClient = walletClient;
    this.contractAddress = contractAddress;

    this.readContract = getContract({
      address: contractAddress,
      abi: executionEngineAbi,
      client: publicClient,
    });
  }

  // -----------------------------------------------------------------------
  // Write operations
  // -----------------------------------------------------------------------

  /**
   * Submit a new execution request on-chain.
   *
   * Calls `requestExecution` on the ExecutionEngine contract. The tip is
   * sent as `msg.value` and must be >= MIN_TIP (0.0001 ETH).
   *
   * @returns The transaction hash
   */
  async submitRequest(options: SubmitRequestOptions): Promise<Hex> {
    const wallet = this.requireWallet();

    const hash = await wallet.writeContract({
      address: this.contractAddress,
      abi: executionEngineAbi,
      functionName: 'requestExecution',
      args: [
        options.imageId,
        options.inputDigest,
        options.inputUrl ?? '',
        options.callbackContract ?? '0x0000000000000000000000000000000000000000',
        options.expirationSeconds ?? 0n,
        options.inputType ?? 0,
      ],
      value: parseEther(options.tipEther),
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
    } as any);

    await this.publicClient.waitForTransactionReceipt({ hash });
    return hash;
  }

  /**
   * Claim an execution request as a prover.
   *
   * Calls `claimExecution(requestId)`. The caller becomes the assigned
   * prover and has CLAIM_WINDOW (10 min) to submit a proof.
   *
   * @returns The transaction hash
   */
  async claimJob(requestId: bigint): Promise<Hex> {
    const wallet = this.requireWallet();

    const hash = await wallet.writeContract({
      address: this.contractAddress,
      abi: executionEngineAbi,
      functionName: 'claimExecution',
      args: [requestId],
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
    } as any);

    await this.publicClient.waitForTransactionReceipt({ hash });
    return hash;
  }

  /**
   * Submit a proof for a claimed execution request.
   *
   * Calls `submitProof(requestId, seal, journal)`. The contract verifies
   * the proof on-chain, pays the prover, and optionally triggers a
   * callback on the requester's contract.
   *
   * @param requestId - The request to fulfill
   * @param seal - The proof seal bytes (risc0 or Remainder)
   * @param journal - The public outputs (journal)
   * @returns The transaction hash
   */
  async submitResult(requestId: bigint, seal: Hex, journal: Hex): Promise<Hex> {
    const wallet = this.requireWallet();

    const hash = await wallet.writeContract({
      address: this.contractAddress,
      abi: executionEngineAbi,
      functionName: 'submitProof',
      args: [requestId, seal, journal],
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
    } as any);

    await this.publicClient.waitForTransactionReceipt({ hash });
    return hash;
  }

  /**
   * Cancel a pending execution request (requester only).
   *
   * Calls `cancelExecution(requestId)`. The tip is refunded to the
   * original requester.
   *
   * @returns The transaction hash
   */
  async cancelRequest(requestId: bigint): Promise<Hex> {
    const wallet = this.requireWallet();

    const hash = await wallet.writeContract({
      address: this.contractAddress,
      abi: executionEngineAbi,
      functionName: 'cancelExecution',
      args: [requestId],
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
    } as any);

    await this.publicClient.waitForTransactionReceipt({ hash });
    return hash;
  }

  /**
   * Dispute / reclaim an expired claim on a request.
   *
   * If a prover claimed a request but failed to submit a proof within
   * the CLAIM_WINDOW, any prover can call `claimExecution` again to
   * reclaim it. This is the "dispute" mechanism: the original claim
   * expires and a new prover takes over.
   *
   * @returns The transaction hash
   */
  async disputeResult(requestId: bigint): Promise<Hex> {
    // The contract's dispute mechanism is reclaiming an expired claim.
    // claimExecution handles this: if status == Claimed and deadline
    // passed, it emits ClaimExpired and allows reclaim.
    return this.claimJob(requestId);
  }

  // -----------------------------------------------------------------------
  // Read operations
  // -----------------------------------------------------------------------

  /**
   * Get the on-chain status of an execution request.
   *
   * @returns The {@link OnChainRequestStatus} enum value
   */
  async getRequestStatus(requestId: bigint): Promise<OnChainRequestStatus> {
    const req = await this.getRequest(requestId);
    return req.status;
  }

  /**
   * Fetch the full on-chain ExecutionRequest struct.
   */
  async getRequest(requestId: bigint): Promise<OnChainExecutionRequest> {
    const result = await this.publicClient.readContract({
      address: this.contractAddress,
      abi: executionEngineAbi,
      functionName: 'getRequest',
      args: [requestId],
    });

    // The ABI returns a tuple. viem decodes it as an object with named fields.
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const r = result as any;
    return {
      id: r.id,
      imageId: r.imageId,
      inputDigest: r.inputDigest,
      requester: r.requester,
      createdAt: Number(r.createdAt),
      expiresAt: Number(r.expiresAt),
      callbackContract: r.callbackContract,
      status: Number(r.status) as OnChainRequestStatus,
      claimedBy: r.claimedBy,
      claimedAt: Number(r.claimedAt),
      claimDeadline: Number(r.claimDeadline),
      tip: r.tip,
      maxTip: r.maxTip,
    };
  }

  /**
   * Get the current tip for a request (decreases over time).
   */
  async getCurrentTip(requestId: bigint): Promise<bigint> {
    return this.publicClient.readContract({
      address: this.contractAddress,
      abi: executionEngineAbi,
      functionName: 'getCurrentTip',
      args: [requestId],
    }) as Promise<bigint>;
  }

  /**
   * Get pending request IDs (for provers to find work).
   */
  async getPendingRequests(offset: bigint = 0n, limit: bigint = 20n): Promise<bigint[]> {
    const result = await this.publicClient.readContract({
      address: this.contractAddress,
      abi: executionEngineAbi,
      functionName: 'getPendingRequests',
      args: [offset, limit],
    });
    return result as bigint[];
  }

  /**
   * Get prover statistics (completed count + earnings).
   */
  async getProverStats(prover: Hex): Promise<OnChainProverStats> {
    const result = await this.publicClient.readContract({
      address: this.contractAddress,
      abi: executionEngineAbi,
      functionName: 'getProverStats',
      args: [prover],
    });
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const r = result as any;
    return {
      completed: r[0] ?? r.completed,
      earnings: r[1] ?? r.earnings,
    };
  }

  /**
   * Get the next request ID that will be assigned.
   */
  async getNextRequestId(): Promise<bigint> {
    return this.publicClient.readContract({
      address: this.contractAddress,
      abi: executionEngineAbi,
      functionName: 'nextRequestId',
    }) as Promise<bigint>;
  }

  /**
   * Check whether the contract is paused.
   */
  async isPaused(): Promise<boolean> {
    return this.publicClient.readContract({
      address: this.contractAddress,
      abi: executionEngineAbi,
      functionName: 'paused',
    }) as Promise<boolean>;
  }

  /**
   * Get the contract owner address.
   */
  async owner(): Promise<Hex> {
    return this.publicClient.readContract({
      address: this.contractAddress,
      abi: executionEngineAbi,
      functionName: 'owner',
    }) as Promise<Hex>;
  }

  /**
   * Get the protocol fee in basis points.
   */
  async getProtocolFeeBps(): Promise<bigint> {
    return this.publicClient.readContract({
      address: this.contractAddress,
      abi: executionEngineAbi,
      functionName: 'protocolFeeBps',
    }) as Promise<bigint>;
  }

  /**
   * Get the fee recipient address.
   */
  async getFeeRecipient(): Promise<Hex> {
    return this.publicClient.readContract({
      address: this.contractAddress,
      abi: executionEngineAbi,
      functionName: 'feeRecipient',
    }) as Promise<Hex>;
  }

  // -----------------------------------------------------------------------
  // Admin operations (owner only)
  // -----------------------------------------------------------------------

  /** Pause the contract (owner only). */
  async pause(): Promise<Hex> {
    return this.send('pause', []);
  }

  /** Unpause the contract (owner only). */
  async unpause(): Promise<Hex> {
    return this.send('unpause', []);
  }

  /** Set the protocol fee in basis points (owner only, max 1000 = 10%). */
  async setProtocolFee(feeBps: bigint): Promise<Hex> {
    return this.send('setProtocolFee', [feeBps]);
  }

  /** Set the fee recipient address (owner only). */
  async setFeeRecipient(recipient: Hex): Promise<Hex> {
    return this.send('setFeeRecipient', [recipient]);
  }

  /** Set or update the reputation contract (owner only). */
  async setReputation(reputationAddress: Hex): Promise<Hex> {
    return this.send('setReputation', [reputationAddress]);
  }

  /** Initiate ownership transfer (owner only, Ownable2Step). */
  async transferOwnership(newOwner: Hex): Promise<Hex> {
    return this.send('transferOwnership', [newOwner]);
  }

  /** Accept pending ownership transfer (new owner only). */
  async acceptOwnership(): Promise<Hex> {
    return this.send('acceptOwnership', []);
  }

  // -----------------------------------------------------------------------
  // Internal helpers
  // -----------------------------------------------------------------------

  private requireWallet(): WalletClient {
    if (!this.walletClient) {
      throw new Error(
        'ExecutionEngineClient: walletClient is required for write operations',
      );
    }
    return this.walletClient;
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  private async send(functionName: string, args: any[], value?: bigint): Promise<Hex> {
    const wallet = this.requireWallet();

    const hash = await wallet.writeContract({
      address: this.contractAddress,
      abi: executionEngineAbi,
      functionName,
      args,
      ...(value !== undefined ? { value } : {}),
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
    } as any);

    await this.publicClient.waitForTransactionReceipt({ hash });
    return hash;
  }
}

File: sdk/python/worldzk/execution_engine.py
"""
ExecutionEngine SDK client for on-chain verifiable computation lifecycle.

Requires the ``web3`` optional dependency: ``pip install worldzk[web3]``

Example usage::

    from worldzk.execution_engine import ExecutionEngineClient

    client = ExecutionEngineClient(
        rpc_url="http://localhost:8545",
        contract_address="0x...",
        private_key="0x...",
    )

    # Submit a request
    request_id = client.submit_request(
        image_id="0x" + "ab" * 32,
        input_hash="0x" + "cd" * 32,
        input_url="https://example.com/inputs.json",
        tip_wei=100000000000000,  # 0.0001 ETH
    )

    # Prover claims the job
    client.claim_job(request_id)

    # Prover submits proof
    client.submit_result(request_id, seal=b"...", journal=b"...")

    # Query status
    status = client.get_request_status(request_id)
    print(f"Status: {status.status}")
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import IntEnum
from typing import Any, List, Optional

from eth_account import Account
from eth_account.signers.local import LocalAccount
from web3 import Web3
from web3.contract import Contract


class RequestStatus(IntEnum):
    """Execution request status matching the on-chain enum."""

    PENDING = 0
    CLAIMED = 1
    COMPLETED = 2
    EXPIRED = 3
    CANCELLED = 4


@dataclass
class ExecutionRequestInfo:
    """On-chain execution request data."""

    id: int
    image_id: bytes
    input_digest: bytes
    requester: str
    created_at: int
    expires_at: int
    callback_contract: str
    status: RequestStatus
    claimed_by: str
    claimed_at: int
    claim_deadline: int
    tip: int
    max_tip: int

    @property
    def is_pending(self) -> bool:
        """Check if request is pending."""
        return self.status == RequestStatus.PENDING

    @property
    def is_claimed(self) -> bool:
        """Check if request has been claimed."""
        return self.status == RequestStatus.CLAIMED

    @property
    def is_completed(self) -> bool:
        """Check if request is completed."""
        return self.status == RequestStatus.COMPLETED

    @property
    def is_expired(self) -> bool:
        """Check if request has expired."""
        return self.status == RequestStatus.EXPIRED

    @property
    def is_cancelled(self) -> bool:
        """Check if request was cancelled."""
        return self.status == RequestStatus.CANCELLED

    @property
    def is_terminal(self) -> bool:
        """Check if request is in a terminal state."""
        return self.status in {
            RequestStatus.COMPLETED,
            RequestStatus.EXPIRED,
            RequestStatus.CANCELLED,
        }


EXECUTION_ENGINE_ABI: list[dict[str, Any]] = [
    # Events
    {
        "type": "event",
        "name": "ExecutionRequested",
        "inputs": [
            {"name": "requestId", "type": "uint256", "indexed": True},
            {"name": "requester", "type": "address", "indexed": True},
            {"name": "imageId", "type": "bytes32", "indexed": True},
            {"name": "inputDigest", "type": "bytes32", "indexed": False},
            {"name": "inputUrl", "type": "string", "indexed": False},
            {"name": "inputType", "type": "uint8", "indexed": False},
            {"name": "tip", "type": "uint256", "indexed": False},
            {"name": "expiresAt", "type": "uint256", "indexed": False},
        ],
        "anonymous": False,
    },
    {
        "type": "event",
        "name": "ExecutionClaimed",
        "inputs": [
            {"name": "requestId", "type": "uint256", "indexed": True},
            {"name": "prover", "type": "address", "indexed": True},
            {"name": "claimDeadline", "type": "uint256", "indexed": False},
        ],
        "anonymous": False,
    },
    {
        "type": "event",
        "name": "ExecutionCompleted",
        "inputs": [
            {"name": "requestId", "type": "uint256", "indexed": True},
            {"name": "prover", "type": "address", "indexed": True},
            {"name": "journalDigest", "type": "bytes32", "indexed": False},
            {"name": "payout", "type": "uint256", "indexed": False},
        ],
        "anonymous": False,
    },
    {
        "type": "event",
        "name": "ExecutionExpired",
        "inputs": [
            {"name": "requestId", "type": "uint256", "indexed": True},
        ],
        "anonymous": False,
    },
    {
        "type": "event",
        "name": "ExecutionCancelled",
        "inputs": [
            {"name": "requestId", "type": "uint256", "indexed": True},
        ],
        "anonymous": False,
    },
    {
        "type": "event",
        "name": "ClaimExpired",
        "inputs": [
            {"name": "requestId", "type": "uint256", "indexed": True},
            {"name": "prover", "type": "address", "indexed": True},
        ],
        "anonymous": False,
    },
    {
        "type": "event",
        "name": "CallbackFailed",
        "inputs": [
            {"name": "requestId", "type": "uint256", "indexed": True},
            {"name": "callbackContract", "type": "address", "indexed": True},
        ],
        "anonymous": False,
    },
    # Functions
    {
        "type": "function",
        "name": "requestExecution",
        "inputs": [
            {"name": "imageId", "type": "bytes32"},
            {"name": "inputDigest", "type": "bytes32"},
            {"name": "inputUrl", "type": "string"},
            {"name": "callbackContract", "type": "address"},
            {"name": "expirationSeconds", "type": "uint256"},
            {"name": "inputType", "type": "uint8"},
        ],
        "outputs": [{"name": "requestId", "type": "uint256"}],
        "stateMutability": "payable",
    },
    {
        "type": "function",
        "name": "cancelExecution",
        "inputs": [{"name": "requestId", "type": "uint256"}],
        "outputs": [],
        "stateMutability": "nonpayable",
    },
    {
        "type": "function",
        "name": "claimExecution",
        "inputs": [{"name": "requestId", "type": "uint256"}],
        "outputs": [],
        "stateMutability": "nonpayable",
    },
    {
        "type": "function",
        "name": "submitProof",
        "inputs": [
            {"name": "requestId", "type": "uint256"},
            {"name": "seal", "type": "bytes"},
            {"name": "journal", "type": "bytes"},
        ],
        "outputs": [],
        "stateMutability": "nonpayable",
    },
    {
        "type": "function",
        "name": "getRequest",
        "inputs": [{"name": "requestId", "type": "uint256"}],
        "outputs": [
            {
                "name": "",
                "type": "tuple",
                "components": [
                    {"name": "id", "type": "uint256"},
                    {"name": "imageId", "type": "bytes32"},
                    {"name": "inputDigest", "type": "bytes32"},
                    {"name": "requester", "type": "address"},
                    {"name": "createdAt", "type": "uint48"},
                    {"name": "expiresAt", "type": "uint48"},
                    {"name": "callbackContract", "type": "address"},
                    {"name": "status", "type": "uint8"},
                    {"name": "claimedBy", "type": "address"},
                    {"name": "claimedAt", "type": "uint48"},
                    {"name": "claimDeadline", "type": "uint48"},
                    {"name": "tip", "type": "uint256"},
                    {"name": "maxTip", "type": "uint256"},
                ],
            }
        ],
        "stateMutability": "view",
    },
    {
        "type": "function",
        "name": "getCurrentTip",
        "inputs": [{"name": "requestId", "type": "uint256"}],
        "outputs": [{"name": "", "type": "uint256"}],
        "stateMutability": "view",
    },
    {
        "type": "function",
        "name": "getPendingRequests",
        "inputs": [
            {"name": "offset", "type": "uint256"},
            {"name": "limit", "type": "uint256"},
        ],
        "outputs": [{"name": "", "type": "uint256[]"}],
        "stateMutability": "view",
    },
    {
        "type": "function",
        "name": "getProverStats",
        "inputs": [{"name": "prover", "type": "address"}],
        "outputs": [
            {"name": "completed", "type": "uint256"},
            {"name": "earnings", "type": "uint256"},
        ],
        "stateMutability": "view",
    },
    {
        "type": "function",
        "name": "nextRequestId",
        "inputs": [],
        "outputs": [{"name": "", "type": "uint256"}],
        "stateMutability": "view",
    },
    {
        "type": "function",
        "name": "protocolFeeBps",
        "inputs": [],
        "outputs": [{"name": "", "type": "uint256"}],
        "stateMutability": "view",
    },
    {
        "type": "function",
        "name": "feeRecipient",
        "inputs": [],
        "outputs": [{"name": "", "type": "address"}],
        "stateMutability": "view",
    },
    {
        "type": "function",
        "name": "owner",
        "inputs": [],
        "outputs": [{"name": "", "type": "address"}],
        "stateMutability": "view",
    },
    {
        "type": "function",
        "name": "paused",
        "inputs": [],
        "outputs": [{"name": "", "type": "bool"}],
        "stateMutability": "view",
    },
    {
        "type": "function",
        "name": "pause",
        "inputs": [],
        "outputs": [],
        "stateMutability": "nonpayable",
    },
    {
        "type": "function",
        "name": "unpause",
        "inputs": [],
        "outputs": [],
        "stateMutability": "nonpayable",
    },
    {
        "type": "function",
        "name": "setProtocolFee",
        "inputs": [{"name": "_feeBps", "type": "uint256"}],
        "outputs": [],
        "stateMutability": "nonpayable",
    },
    {
        "type": "function",
        "name": "setFeeRecipient",
        "inputs": [{"name": "_recipient", "type": "address"}],
        "outputs": [],
        "stateMutability": "nonpayable",
    },
]


def _to_bytes32(value: bytes | str) -> bytes:
    """Convert to 32-byte value."""
    if isinstance(value, str):
        if value.startswith("0x") or value.startswith("0X"):
            value = bytes.fromhex(value[2:])
        else:
            value = bytes.fromhex(value)
    if len(value) != 32:
        raise ValueError(f"Expected 32 bytes, got {len(value)}")
    return value


def _to_bytes(value: bytes | str) -> bytes:
    """Convert hex string or bytes to bytes."""
    if isinstance(value, str):
        if value.startswith("0x") or value.startswith("0X"):
            return bytes.fromhex(value[2:])
        return bytes.fromhex(value)
    return value


# Minimum tip required by the contract (0.0001 ETH)
MIN_TIP_WEI = 100_000_000_000_000


class ExecutionEngineClient:
    """Client for the ExecutionEngine smart contract.

    Provides methods for the full execution lifecycle:
    request submission, job claiming, proof submission, and status queries.

    Can be used in read-only mode (without private_key) for queries,
    or with a private_key for submitting transactions.
    """

    def __init__(
        self,
        rpc_url: str,
        contract_address: str,
        private_key: Optional[str] = None,
        gas_limit: int = 500_000,
    ):
        """Create a new ExecutionEngineClient.

        Args:
            rpc_url: JSON-RPC endpoint URL.
            contract_address: Deployed ExecutionEngine contract address.
            private_key: Hex-encoded private key for signing transactions.
                         If None, only read-only methods are available.
            gas_limit: Maximum gas per transaction (default 500000).
        """
        self._w3 = Web3(Web3.HTTPProvider(rpc_url))
        self._account: Optional[LocalAccount] = None
        if private_key is not None:
            self._account = Account.from_key(private_key)
        self._contract: Contract = self._w3.eth.contract(
            address=Web3.to_checksum_address(contract_address),
            abi=EXECUTION_ENGINE_ABI,
        )
        self._gas_limit = gas_limit

    @property
    def address(self) -> str:
        """The contract address."""
        return self._contract.address

    @property
    def account_address(self) -> Optional[str]:
        """The signer's address, or None if read-only."""
        if self._account is not None:
            return self._account.address
        return None

    def _require_signer(self) -> LocalAccount:
        """Ensure a private key was provided for transaction signing."""
        if self._account is None:
            raise ValueError(
                "Private key required for transactions. "
                "Pass private_key to the constructor."
            )
        return self._account

    # ===================================================================
    # TRANSACTION METHODS
    # ===================================================================

    def submit_request(
        self,
        image_id: bytes | str,
        input_hash: bytes | str,
        input_url: str = "",
        callback_contract: str = "0x0000000000000000000000000000000000000000",
        expiration_seconds: int = 0,
        input_type: int = 0,
        tip_wei: int = MIN_TIP_WEI,
    ) -> int:
        """Submit an execution request.

        Args:
            image_id: The program image ID (bytes32).
            input_hash: Hash of the inputs (bytes32).
            input_url: URL where provers can fetch the inputs.
            callback_contract: Contract to receive results (zero address for none).
            expiration_seconds: Expiration time in seconds (0 = default 1 hour).
            input_type: 0 = public, 1 = private.
            tip_wei: Tip amount in wei (must be >= MIN_TIP = 0.0001 ETH).

        Returns:
            The request ID assigned by the contract.

        Raises:
            ValueError: If tip is below minimum or private key not set.
        """
        if tip_wei < MIN_TIP_WEI:
            raise ValueError(
                f"Tip must be at least {MIN_TIP_WEI} wei (0.0001 ETH), got {tip_wei}"
            )

        receipt = self._send_tx(
            self._contract.functions.requestExecution(
                _to_bytes32(image_id),
                _to_bytes32(input_hash),
                input_url,
                Web3.to_checksum_address(callback_contract),
                expiration_seconds,
                input_type,
            ),
            value=tip_wei,
        )

        # Parse request ID from ExecutionRequested event
        logs = self._contract.events.ExecutionRequested().process_receipt(receipt)
        if logs:
            return logs[0]["args"]["requestId"]

        # Fallback: cannot determine request ID from receipt
        raise RuntimeError("ExecutionRequested event not found in transaction receipt")

    def claim_job(self, request_id: int) -> str:
        """Claim an execution request as a prover.

        Args:
            request_id: The request to claim.

        Returns:
            Transaction hash as hex string.
        """
        receipt = self._send_tx(
            self._contract.functions.claimExecution(request_id)
        )
        return self._tx_hash_hex(receipt)

    def submit_result(
        self,
        request_id: int,
        seal: bytes | str,
        journal: bytes | str,
    ) -> str:
        """Submit proof for a claimed execution request.

        Args:
            request_id: The request ID.
            seal: The proof seal bytes.
            journal: The public outputs (journal) bytes.

        Returns:
            Transaction hash as hex string.
        """
        receipt = self._send_tx(
            self._contract.functions.submitProof(
                request_id,
                _to_bytes(seal),
                _to_bytes(journal),
            )
        )
        return self._tx_hash_hex(receipt)

    def cancel_request(self, request_id: int) -> str:
        """Cancel a pending execution request and refund the tip.

        Only the original requester can cancel.

        Args:
            request_id: The request to cancel.

        Returns:
            Transaction hash as hex string.
        """
        receipt = self._send_tx(
            self._contract.functions.cancelExecution(request_id)
        )
        return self._tx_hash_hex(receipt)

    # ===================================================================
    # READ-ONLY METHODS
    # ===================================================================

    def get_request_status(self, request_id: int) -> ExecutionRequestInfo:
        """Get on-chain request details.

        Args:
            request_id: The request to query.

        Returns:
            ExecutionRequestInfo with full request details.
        """
        raw = self._contract.functions.getRequest(request_id).call()
        return ExecutionRequestInfo(
            id=raw[0],
            image_id=raw[1],
            input_digest=raw[2],
            requester=raw[3],
            created_at=raw[4],
            expires_at=raw[5],
            callback_contract=raw[6],
            status=RequestStatus(raw[7]),
            claimed_by=raw[8],
            claimed_at=raw[9],
            claim_deadline=raw[10],
            tip=raw[11],
            max_tip=raw[12],
        )

    def get_current_tip(self, request_id: int) -> int:
        """Get the current tip for a request (decreases over time via decay).

        Args:
            request_id: The request to query.

        Returns:
            Current tip amount in wei.
        """
        return self._contract.functions.getCurrentTip(request_id).call()

    def get_pending_requests(
        self, offset: int = 0, limit: int = 20
    ) -> List[int]:
        """Get pending request IDs for provers to find work.

        Args:
            offset: Number of matching requests to skip.
            limit: Maximum number of request IDs to return.

        Returns:
            List of request IDs that are currently pending.
        """
        return self._contract.functions.getPendingRequests(offset, limit).call()

    def get_prover_stats(self, prover_address: str) -> tuple[int, int]:
        """Get prover statistics.

        Args:
            prover_address: The prover address to query.

        Returns:
            Tuple of (completed_count, total_earnings_wei).
        """
        result = self._contract.functions.getProverStats(
            Web3.to_checksum_address(prover_address)
        ).call()
        return (result[0], result[1])

    def get_next_request_id(self) -> int:
        """Get the next request ID that will be assigned."""
        return self._contract.functions.nextRequestId().call()

    def get_protocol_fee_bps(self) -> int:
        """Get the protocol fee in basis points."""
        return self._contract.functions.protocolFeeBps().call()

    def get_fee_recipient(self) -> str:
        """Get the protocol fee recipient address."""
        return self._contract.functions.feeRecipient().call()

    def owner(self) -> str:
        """Get the contract owner address."""
        return self._contract.functions.owner().call()

    def paused(self) -> bool:
        """Check if the contract is paused."""
        return self._contract.functions.paused().call()

    # ===================================================================
    # DISPUTE METHODS (maps to task requirement)
    # ===================================================================

    def dispute_result(self, request_id: int) -> str:
        """Initiate a dispute on a completed request.

        Note: The ExecutionEngine contract uses a proof-based verification
        model (submit proof at completion time), not a post-hoc dispute
        model. This method cancels a pending request as the closest
        available on-chain dispute mechanism. For TEE-style challenge/dispute
        flows, use TEEVerifier instead.

        Args:
            request_id: The request to dispute/cancel.

        Returns:
            Transaction hash as hex string.
        """
        return self.cancel_request(request_id)

    # ===================================================================
    # INTERNAL
    # ===================================================================

    def _send_tx(self, fn: Any, value: int = 0) -> Any:
        """Build, sign, and send a transaction."""
        account = self._require_signer()
        tx = fn.build_transaction({
            "from": account.address,
            "nonce": self._w3.eth.get_transaction_count(account.address),
            "gas": self._gas_limit,
            "gasPrice": self._w3.eth.gas_price,
            "value": value,
        })
        try:
            estimated = self._w3.eth.estimate_gas(tx)
            tx["gas"] = int(estimated * 1.1)
        except Exception:
            pass
        signed = account.sign_transaction(tx)
        tx_hash = self._w3.eth.send_raw_transaction(signed.raw_transaction)
        return self._w3.eth.wait_for_transaction_receipt(tx_hash)

    @staticmethod
    def _tx_hash_hex(receipt: Any) -> str:
        """Extract tx hash as hex string."""
        h = receipt["transactionHash"]
        if isinstance(h, bytes):
            return "0x" + h.hex()
        return str(h)
